Implement Legacy Widget design iterations

noisysocks opened the pull request:
Closes https://github.com/WordPress/gutenberg/issues/26179.
Closes https://github.com/WordPress/gutenberg/issues/28671.
Closes https://github.com/WordPress/gutenberg/issues/25960.
Closes https://github.com/WordPress/gutenberg/issues/28995.

Implements the new Legacy Widget flow in https://github.com/WordPress/gutenberg/issues/26179.

- The widget preview now displays by default.
- The widget form now displays when the block is selected or when user presses <kbd>ESC</kbd>.
- There is now a "No preview available" message when the widget has no preview.

https://user-images.githubusercontent.com/612155/115821121-8b32e800-a445-11eb-9035-b8f1687b55c1.mp4

noisysocks committed:
Legacy Widget: Show Preview by default and add Apply and Cancel buttons

noisysocks committed:
Legacy Widget: Add 'No preview available' placeholder

noisysocks committed:
Legacy Widget: Update form and 'No preview available' placeholder styling

noisysocks commented:
@critterverse: This is still a work in progress but let me know what you think! Also please feel free to push any CSS changes you'd like to make. 

What do you think about the _Apply_ / _Cancel_ flow? I have two concerns:

- The way it interacts with _Undo_ / _Redo_ isn't 100% clear. Is it correct that pressing <kbd>Cmd</kbd>+<kbd>Z</kbd> should undo all of the changes that last were made when the user clicked _Apply_? 
- It makes the Legacy Widget block behave differently to regular blocks. To update a Quote block, one doesn't make their changes and then click _Apply_. One simply makes their changes. Why must Legacy Widget be different?

noisysocks committed:
Add border to Legacy Widget form

noisysocks committed:
Legacy Widget: Move useForm() up a level

noisysocks committed:
Legacy Widget: Avoid unmounting preview so that transition is smooth

noisysocks committed:
Hide NoPreview when not selected

noisysocks committed:
Don't show hover effect when Legacy Widget is loading

noisysocks committed:
Fix jumping text in NoPreview

noisysocks committed:
Improve placeholder state

noisysocks committed:
Show placeholder while preview iframe is loading so there's no jankiness

noisysocks commented:
I made some more progress on this today and have updated the video in the PR's description. I still don't really like the _Apply_ / _Cancel_ flow but I think everything else is pretty smooth! 🙂

kevin940726 commented on packages/block-library/src/legacy-widget/edit/preview.js:
```diff
@@ -1,53 +1,92 @@
+				title={ title }
+				height={ height ?? defaultHeight }
+				onLoad={ () => {
+					setHeight( ref.current.contentDocument.body.scrollHeight );
```
We could probably just use `event.target` so that we don't have to attach a ref.

kevin940726 replied:
Would reloading the iframe re-trigger this event? Or when something like changing the widget type?

noisysocks replied:
I did this! And yes `onLoad` will trigger when the `src` changes.

kevin940726 commented on packages/block-library/src/legacy-widget/edit/preview.js:
```diff
@@ -1,53 +1,92 @@
+function PreviewIframe( {
+	className,
+	title,
+	height: defaultHeight = 10,
```
When will this `defaultHeight` ever be used?

noisysocks committed:
Fix Cancel not clearing out changes

noisysocks committed:
Remove Apply / Cancel flow

critterverse commented:
Hi all, took this PR for a quick spin and have a few notes below!

**Re: Apply/Cancel flow**

>It makes the Legacy Widget block behave differently to regular blocks. To update a Quote block, one doesn't make their changes and then click Apply. One simply makes their changes. Why must Legacy Widget be different?

Great question, @noisysocks. The Legacy Widget has the added complexity of switching between a preview and an editing interface, so I can see why the behavior would be different from a regular block. That being said and now that I’m playing with the PR — removing these actions _does_ feel like it could potentially help simplify the UI and editing flow.

One issue I can see with removing the Apply/Cancel part of the editing flow is when it comes to Legacy Widgets without a preview — there’s not really a good way to confirm that your changes have been applied since you don’t get the visual confirmation of seeing the preview updated.

Would love to hear from @shaunandrews on this, who may have other watch-outs that I'm overlooking.

>The way it interacts with Undo/Redo isn't 100% clear. Is it correct that pressing Cmd+Z should undo all of the changes that last were made when the user clicked Apply?

Yup — if this behavior remains, Cmd+Z should undo all of the changes that were made when the user clicked _Apply_. Didn't test this extensively but appears to be working correctly in the PR 👍

**Styling**

Adding a few top-level notes that stood out for the Legacy Widgets (happy to help make these changes but also don’t want to hold anything up!):

![Screen Shot 2021-04-20 at 10 56 07 AM](https://user-images.githubusercontent.com/77359364/115419835-a3e29a00-a1c8-11eb-9f6a-e9578bf996f2.png)

- [x] Needs 12px space between labels and input fields (for example “Title” and the text field beneath it)
- [x] Block toolbar can align to the left edge of the block
- [x] Depending on what we decide with the Apply/Cancel behavior — I think it should be moved after the “Move to Widget area” action in the toolbar so that it appears last (before the ellipses menu)
- [x] The block description that appears in the inspector for Legacy Widgets feels like it could use some attention — maybe just left aligning the text and removing the “Widget:” from the 3rd line (so it would just say “Recent Posts” in the example below)

![Screen Shot 2021-04-20 at 10 58 56 AM](https://user-images.githubusercontent.com/77359364/115420247-fb810580-a1c8-11eb-9fd9-3ad5d13100f6.png)

noisysocks commented:
Thanks for the feedback @critterverse! I'll get to work on fixing those styling issues.

Here's how this looks with a _Done editing_ button instead of _Apply_ / _Cancel_. You can play with it yourself by checking out the `try/legacy-widget-design-with-no-apply-cancel` branch which I just pushed.

https://user-images.githubusercontent.com/612155/115483106-7b73a200-a293-11eb-9638-4594d400940a.mp4

noisysocks committed:
Add margin between label and input for most legacy widgets

noisysocks commented:
> * Block toolbar can align to the left edge of the block

I ticketed this one separately in https://github.com/WordPress/gutenberg/issues/31021 as it is already an issue in `trunk` and the fix requires moving a few unrelated things around.

noisysocks committed:
Move Apply & Cancel buttons to 'other' group

noisysocks committed:
Tidy up LegacyWidgetInspectorCard

noisysocks committed:
Fix PHP tests

kevin940726 commented on packages/block-library/src/legacy-widget/edit/form.js:
```diff
@@ -36,127 +20,37 @@ export default function Form( { id, idBase, instance, setInstance } ) {
 			onSave={ setFormData }
+			// Force a remount when the widget's form HTML changes. This clears
+			// out any mutations to the DOM that widget scripts have made.
+			key={ content.key }
```
I think using `key` to force a remount is fine, but I don't like using an arbitrary `key` variable for it. Could we just use `html` as the key? It might impact performance, but maybe it's insignificant.

noisysocks replied:
We can't because then the _Cancel_ flow won't work.

To illustrate:

1. User has a text widget. The form HTML (`content.html`) is `"<textarea>Foo</textarea>"`.
2. User appends "Bar". `useForm` detects change to `formData`, sends XHR request. It updates `pendingInstance` to `{ "content": "FooBar" }` but does **not** update form HTML (`content.html`) because we do not want to cause a bad UX by remounting the form after user input.
3. User presses _Cancel_. `pendingInstance` is reset to `instance` which is `{ "content": "Foo" }`. `useForm` detects this, sends XHR request, and sets form HTML (`content.html`) back to `"<textarea>Foo</textarea>"`.
4. Form component would **not** remount because the key has not changed (`"<textarea>Foo</textarea>" === "<textarea>Foo</textarea>"`).

Happy to change this if you can think of a better approach. An explicit `key` was the best I could come up with 😅. Maybe it should be called `content.version`.

kevin940726 replied:
Do we still have a cancel flow now? 😅

noisysocks replied:
Right you are! Fixed in ef778996f0c4b032a6a3aaa09ffd76389dd18659.

critterverse commented:
Sorry for the confusion, just to clarify on my last comment — I thought the alternative suggestion was to potentially remove the confirmation step entirely, i.e. having the click-away be the _Apply_ action... but sounds like I was misunderstanding on that 😅

Re: confirmation step, I would still vote for _Apply/Cancel_ over _Done Editing_ because it already exists in other unique "editing contexts" like the toolbar for the Image block (and could potentially be extended to the Cover block soon as more editing capabilities are added there). 

Styling updates look great! Something that may have regressed in the latest push or maybe I'm just noticing is that the tab in the inspector should switch over from the Widget Area tab to the Block tab when the block is selected:

![Screen Shot 2021-04-21 at 1 25 14 PM](https://user-images.githubusercontent.com/77359364/115597439-44a78700-a2a7-11eb-9e98-37806e70a6dc.png)

noisysocks commented:
> Sorry for the confusion, just to clarify on my last comment — I thought the alternative suggestion was to potentially remove the confirmation step entirely, i.e. having the click-away be the _Apply_ action... but sounds like I was misunderstanding on that 😅

You thought correct! In https://github.com/WordPress/gutenberg/pull/30889#issuecomment-823706543, if you click away, the changes will be applied. I forgot to demo it 😄. The _Done editing_ button is just a convenience button that unselects the block. We could get rid of it.

noisysocks committed:
Fix block inspector not opening when block is selected

noisysocks commented:
> Styling updates look great! Something that may have regressed in the latest push or maybe I'm just noticing is that the tab in the inspector should switch over from the Widget Area tab to the Block tab when the block is selected:

Good catch! Looks like this was already broken in `trunk` but it's an easy fix so I added it to this PR in 35611a07fc.

noisysocks committed:
Fix multisite PHP unit tests

kevin940726 commented on packages/edit-widgets/src/components/sidebar/index.js:
```diff
@@ -118,7 +118,7 @@ export default function Sidebar() {
 		) {
 			enableComplementaryArea(
-				editWidgetsStore,
+				'core/edit-widgets',
```
I believe a better idea is to use `.name`.

```diff
-'core/edit-widgets',
+editWidgetsStore.name,
```

noisysocks replied:
In this context `'core/edit-widgets'` is the name of the complementary area and not the name of the store.

critterverse commented:
>You thought correct! In https://github.com/WordPress/gutenberg/pull/30889#issuecomment-823706543, if you click away, the changes will be applied. I forgot to demo it 😄. The Done editing button is just a convenience button that unselects the block. We could get rid of it.

Ok, good to know! I really like this idea :) 

I think we can get rid of the _Done editing_ confirmation step entirely — just want to make sure that the editing flow still works smoothly in that case from an accessibility standpoint. 

One solution could be to use the `Escape` key to switch to Navigation/Select mode _and_ switch from the editing interface to the preview. So in Navigation mode, you would only ever see the preview. Does this feel like something that could work?

noisysocks commented:
Alright I've removed the toolbar buttons, made it so that the preview shows when in Navigation mode, and updated the video in the PR's description. I'm pretty happy with how this looks and feels!

All that's left is to fix the Legacy Widget E2E test which is proving incredibly difficult because it's hard in Puppeteer to tell when an iframe updates. I may have to just disable the test. In the meantime I'll mark this as ready for review.

critterverse commented:
Well done @noisysocks! The simplicity of the toggle between the editing/preview modes using `Escape` and `Enter` feels 🤩

noisysocks committed:
Update Legacy Widget E2E test

noisysocks commented:
Tests pass! Didn't have to disable the one that was causing me grief. This is ready for review.

draganescu commented:
I have tested this, it works as described, but:

1. It was very slow in Firefox:
- clicking on a widget freezes the UI for a bit only then showing the edit form
- in the edit form typing in inputs froze at the 1st character, then becoming smooth, no matter which form, always the 1st character (Firefox)
- In Chrome the speed was fine
2. Some widgets (for example Site Origin Accordion) once de-focused are invisible as previews, one has to hunt for them on the white canvas.
3. In the Customizier where I was able to setup a Site Origin accordion, because the form interaction works, the legacy widget's preview is as tall as the widget's form leading to a lot of whitespace.

noisysocks commented:
@draganescu: Hm, I'm not seeing the performance issues in Firefox. Do you have an extension which is impacting performance? Or is DevTools open?

noisysocks committed:
Remove content.key

noisysocks commented:
> Some widgets (for example Site Origin Accordion) once de-focused are invisible as previews, one has to hunt for them on the white canvas.

~Hm, this is hard to fix. We display the "No preview available." placeholder when the widget's `widget()` function outputs nothing or only whitespace. The SiteOrigin Accordion widget outputs some empty `<div>` markup which is why we don't display "No preview available." The built-in Text widget has a similar problem if you clear out all of the text content.~

~We should document this behaviour and encourage widgets to render nothing instead of empty markup.~

~Aside from that, not sure what we can do. Perhaps display a header containing the widget's title above the preview at all times? What do you think @critterverse?~

**edit:** Never mind, see my comment below.

noisysocks committed:
Use innerText to determine if preview is empty

noisysocks commented:
Actually, never mind, we can use `innerText !== ''` to determine if the preview HTML has actual content in it. Fixed in 4491cc5782abb9ccaf5e509290a77360428a7cea.
